### PR TITLE
Jwe Json serialization support

### PR DESCRIPTION
**Pull request by gjermystreeva**
- Addresses #75 (support JWE JSON serialization too)
- Have not (intentionally) broken any functionality or introduced regressions. (Other that net40 removal from project which was broken already?)
- Supports "General JWE JSON Serialization" and "Flattened JWE JSON Serialization"
- Choice of general / flattened on encrypt made by library based on whether >1 recipient where SerialzationMode.smJson is specified.
- Handles either general or flattened on decrypt where SerialzationMode.smJson is specified.
- IKeyManagement implementations have to now support a WrapKey method. (Using an existing CEK). For a couple that is not valid (re-using an existing CEK) - unittests for that.
- Have added multiple framework targets for the unit test project.
- Have added GitHub actions for unit testing on a matrix of frameworks + OSes. Ideally before merge should fix or mark as expected failures where there is not support / failing tests. Most failures are pre-existing but now exposed without having to setup the environments ones self.
- I had changed the implementation in JWT.cs to call into new JWE.cs - but backed-out for now as would remove net461 support, however I'm not convinced how good the support left in master is or should be for net461. I can clean things up if can remove net461.
- Oh - no tests yet for passing a custom JsonMapper - I'm not 100% sure whether need to or should support that.

### CODE REVIEW

File: jose-jwt/jose-jwt.csproj [hunk 1]
```diff
@@ -93,7 +93,8 @@ RFC 7797:
     <NeutralLanguage>en-US</NeutralLanguage>
     <VersionPrefix>3.0.0</VersionPrefix>
     <Authors>Dmitry Vsekhvalnov</Authors>
-    <TargetFrameworks>net40;net461;netstandard1.4;netstandard2.1</TargetFrameworks>
+    <!--<TargetFrameworks>net40;net461;netstandard1.4;netstandard2.1</TargetFrameworks>-->
+	<TargetFrameworks>net461;netstandard1.4;netstandard2.1</TargetFrameworks>
```
Comment (gjermystreeva) [inline jose-jwt/jose-jwt.csproj:97]: - net40 should definitely go (imo)
- net461... if can delete that also would mean more clean-up possible

File: jose-jwt/jwe/JweJson.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
+﻿#if NETSTANDARD
+namespace Jose.jwe
+{
+    using Newtonsoft.Json;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Internal class used as helper deserialize Json Serialized JWE
+    /// </summary>
+    internal class JweJson
```
Comment (gjermystreeva) [inline jose-jwt/jwe/JweJson.cs:10]: Helper (internal) class for JSON deserialization. Not actually use for serialization, Could (maybe) be removed if ok to delegate all the (de)serialization to the settings supplied class.

**Comment by dvsekhvalnov** [comment]
Hi @gjermystreeva , 

sorry for delay, i started looking into your PR. Very much appreciate your effort.
Given amount of changes, it will take some time for me to refresh spec in my head, so bear with me :) 

1. Did i capture it correctly, desired interface to use it is:

```c#
string json = Jose.Jwe.Encrypt("top secret information", new [] {new JweRecipient(JweAlgorithm.A256KW, key), new JweRecipient(JweAlgorithm.ECDH_ES, ecKey)}, JweEncryption.A256GCM);
```
?

2. Would it make sense to collapse to existing interface may be? Something like:

```c#
string json = Jose.JWT.Encode("top secret information", new [] {new JweRecipient(JweAlgorithm.A256KW, key), new JweRecipient(JweAlgorithm.ECDH_ES, ecKey)}, JweEncryption.A256GCM); 
```

seems it should fit with another overload (though may be we have too much overloads..).

3. By chance have you tried (or know) it against any other library that supports jwe-json serialization? Compatibility is always important thing.

File: jose-jwt/jwe/JWE.cs [hunk 3]
```diff
@@ -0,0 +1,330 @@
+﻿#if NETSTANDARD
+namespace Jose.jwe
```
Comment (dvsekhvalnov) [inline jose-jwt/jwe/JWE.cs:2]: Probably will be nicer to rename to: `Jose.Jwe`

File: jose-jwt/jwe/JWE.cs [hunk 4]
```diff
@@ -0,0 +1,330 @@
+﻿#if NETSTANDARD
+namespace Jose.jwe
+{
+    using Newtonsoft.Json;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+
+    public class JweRecipient
```
Comment (dvsekhvalnov) [inline jose-jwt/jwe/JWE.cs:10]: and this guy to `Recipient`. 

So outside it all become: 
```c#
using Jose;

new Jwe.Recipient(..);
```

?

File: jose-jwt/jwe/JWE.cs [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+            {
+                joseProtectedHeader["zip"] = settings.CompressionHeader(compression.Value);
+                plaintext = settings.Compression(compression.Value).Compress(plaintext);
+            }
+
+            switch (mode)
+            {
+                case SerializationMode.smCompact:
+                    {
+                        if (recipientsOut.Count != 1)
```
Comment (dvsekhvalnov) [inline jose-jwt/jwe/JWE.cs:121]: Compact mode doesn't seem to make anything different from `Jose.JWT.Encode`, right?
What's the benefit to duplicate it here?

**Comment by gjermystreeva** [comment]
Hi @dvsekhvalnov,

> Hi @gjermystreeva ,
> 
> sorry for delay, i started looking into your PR. Very much appreciate your effort.
> Given amount of changes, it will take some time for me to refresh spec in my head, so bear with me :)
> 
> 1. Did i capture it correctly, desired interface to use it is:
> 
> ```cs
> string json = Jose.Jwe.Encrypt("top secret information", new [] {new JweRecipient(JweAlgorithm.A256KW, key), new JweRecipient(JweAlgorithm.ECDH_ES, ecKey)}, JweEncryption.A256GCM);
> ```
> 

Yes - that is basically correct, though currently you also need to supply the mode parameter (SerializationMode.smJson) if you want Json serialized (which you need in this example as 2 recipients).
JweTest.cs "EncryptDecrypt_ModeGeneralJsonRoundTripMultipleRecipients_ValidRecipientsCanDecrypt" shows a typical call.

> 
> 1. Would it make sense to collapse to existing interface may be? Something like:
> 
> ```cs
> string json = Jose.JWT.Encode("top secret information", new [] {new JweRecipient(JweAlgorithm.A256KW, key), new JweRecipient(JweAlgorithm.ECDH_ES, ecKey)}, JweEncryption.A256GCM); 
> ```
> 
> seems it should fit with another overload (though may be we have too much overloads..).
> 

I didn't want to overload the existing JWT class any further, but preferred a new JWE specific API, similar to the discussion in #75.
Similar to how the API is split for example in https://jwcrypto.readthedocs.io/en/latest/index.html

> 1. By chance have you tried (or know) it against any other library that supports jwe-json serialization? Compatibility is always important thing.

I haven't explicitly got tests yet against other libraries, but I have got unit tests based on examples from [rfc7516](https://tools.ietf.org/html/rfc7516) as well as the useful [rfc7520](https://tools.ietf.org/html/rfc7520) which has extra examples.

I have read through the codebase for Python's jwcrypto library and also [go-jose](https://github.com/square/go-jose) for reference.

(My plan is to do some testing against Python's jwcrypto library also.)

File: jose-jwt/jwe/JWE.cs (hunk 3 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs]: Now done.
Also note class Jwe renamed to class JWE to be consistent with class naming for JWT, and to avoid name clash with the namespace.

File: jose-jwt/jwe/JWE.cs (hunk 4 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs]: Now done

File: jose-jwt/jwe/JWE.cs (hunk 5 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs:121]: I've updated PR so that JWT.EncodeBytes now calls through to JWE.Encrypt, so no duplication and JWT class gets simplified.

File: jose-jwt/jwe/JWE.cs [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        public JweAlgorithm Alg { get; set; }
+
+        public object Key { get; set; }
+
+        internal IDictionary<string, object> PerRecipientHeaders { get; }
+    }
+
+    public enum SerializationMode
+    {
+        smCompact,
```
Comment (dvsekhvalnov) [inline jose-jwt/jwe/JWE.cs:33]: Probably just `Compact` and `Json`? `sm` prefix looks so 1990ish :)

File: .github/workflows/build-test.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
+name: BuildAndTestFunctions
```
Comment (dvsekhvalnov) [inline .github/workflows/build-test.yml:1]: is it working at all? i mean running something?

File: UnitTests/UnitTests.csproj [hunk 8]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <Description>UnitTests Class Library</Description>
     <Authors>Administrator</Authors>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFrameworks>netcoreapp2.1;netcoreapp3.1</TargetFrameworks>
```
Comment (dvsekhvalnov) [inline UnitTests/UnitTests.csproj:6]: m? any reason to return 2.1 in test project?

File: jose-jwt/jose-jwt.csproj [hunk 9]
```diff
@@ -93,7 +93,7 @@ RFC 7797:
     <NeutralLanguage>en-US</NeutralLanguage>
     <VersionPrefix>3.0.0</VersionPrefix>
     <Authors>Dmitry Vsekhvalnov</Authors>
-    <TargetFrameworks>net40;net461;netstandard1.4;netstandard2.1</TargetFrameworks>
+    <TargetFrameworks>netstandard1.4;netstandard2.1</TargetFrameworks>
```
Comment (dvsekhvalnov) [inline jose-jwt/jose-jwt.csproj:96]: let's keep `net40`, don't believe it is part of any .netstandard.

Was it causing any build issues with new changes?

File: jose-jwt/jose-jwt.csproj [hunk 10]
```diff
@@ -144,4 +145,8 @@ RFC 7797:
       <Compile Remove="RsaKey.cs" />
     </ItemGroup>
 
+    <ItemGroup>
```
Comment (dvsekhvalnov) [inline jose-jwt/jose-jwt.csproj:148]: why we do we need it?

File: jose-jwt/jose-jwt.csproj [hunk 11]
```diff
@@ -117,6 +117,7 @@ RFC 7797:
   </ItemGroup>
 
   <ItemGroup Condition=" '$(TargetFramework)' == 'net461' ">
+	<PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
```
Comment (dvsekhvalnov) [inline jose-jwt/jose-jwt.csproj:120]: why do we need it? .net46 should include built-in json support. 
My plan was actually to drop newtonsoft going forward as netcore 3 includes built-in too.

File: jose-jwt/jwe/JWE.cs [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+
+            var ret = new List<IDictionary<string, object>>();
+            foreach(var recipient in parsedJwe.Recipients)
+            {
+                ret.Add(MergeHeaders(protectedHeaders, parsedJwe.UnprotectedHeader, recipient.Header));
+            }
+            return ret;
+        }
+
+        private static IDictionary<string, object> MergeHeaders(params IDictionary<string, object>[] dicts)
```
Comment (dvsekhvalnov) [inline jose-jwt/jwe/JWE.cs:316]: Can we either use existing `Dictionaries.Append()` or move `MergeHeades()` under `Dictionaries` too + add unit tests?

File: jose-jwt/jwe/JWE.cs [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+            settings = GetSettings(settings);
+            IJweAlgorithm _enc = settings.Jwe(enc);
+
+            if (_enc == null)
+            {
+                throw new JoseException(string.Format("Unsupported JWE enc requested: {0}", enc));
+            }
+
+            IDictionary<string, object> joseProtectedHeader = MergeHeaders(
```
Comment (dvsekhvalnov) [inline jose-jwt/jwe/JWE.cs:68]: question here, original implementation do not allow to override `enc` and `alg` header values when passing them as `extraHeaders` param. Is it still maintained? Little bit hard to follow logic here as `MergeHeaders()` called multiple times.

File: .github/workflows/build-test.yml (hunk 7 above)
Comment (gjermystreeva) [reply .github/workflows/build-test.yml:1]: Yes - see https://github.com/streeva/jose-jwt/runs/1645966664?check_suite_focus=true.
Same will run on your parent (dvsekhvalnov/jose-jwt) repo one any merge of the PR,

File: UnitTests/UnitTests.csproj (hunk 8 above)
Comment (gjermystreeva) [reply UnitTests/UnitTests.csproj:6]: Yes - again see https://github.com/streeva/jose-jwt/runs/1645966664?check_suite_focus=true.
The action is run as a matrix so you can see what is failing or passing running against both netcoreapp2.1 and netcoreapp3.1. Also the difference between windows / linux os.
Currently only windows netcoreapp3.1 of those combinations has a clean run - but that is basically the existing reality (I believe).

File: jose-jwt/jose-jwt.csproj (hunk 9 above)
Comment (gjermystreeva) [reply jose-jwt/jose-jwt.csproj]: - I don't have any environment to test net40. I believe it went out of support by MS in 2016... is it really required to maintain support for this on the master branch / new releases? (Could you security patch fix a net40 branch if requried?).
- Note I've done more work testing net461 and have today made further commit to get the net461 unit test project clean. - Also I added 461 as a target of the 'main' unit test project and fixed some existing breakages to see how well the main project runs against 461.
- I think for net40 though I wouldn't get any time allowed to validate.

File: jose-jwt/jose-jwt.csproj (hunk 10 above)
Comment (gjermystreeva) [reply jose-jwt/jose-jwt.csproj]: - System.ValueTuple... I return a tuple from JWE.Decrypt.. you get the plaintext, Jose headers (and aad if present),
- On net461 and netstandard1.4 need to include this as a project dependency. For netstandard2.1 its not needed.
- I've cleaned the csproj to only include for those targets now and committed the change.

File: jose-jwt/jose-jwt.csproj (hunk 11 above)
Comment (gjermystreeva) [reply jose-jwt/jose-jwt.csproj]: Yes - ok... I've now removed that dependency for net461 from the main project and use only the wrapping class now not newtonsoft directly.

File: jose-jwt/jwe/JWE.cs (hunk 6 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs]: Sure - done.

File: jose-jwt/jwe/JWE.cs (hunk 13 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs:66]: Yes - I did already add explicit test for `enc `and `alg `overrides - see `Encrypt_WithNonUniqueHeaderParameterNamesInExtraHeaders_Throws`.
That test covers that case explicitly.
There is a slight change in behaviour though (which test documents), that with new implementation the function throws out on attempt to override `enc `or `alg` (or to specify another header twice in both the per-recipient and extra headers), rather than silently ignore a subset of the passed extra parameters.
I'll leave this one open rather than mark resolved in case you have strong opinion on not throwing for 'bad' arguments,

File: jose-jwt/jwe/JWE.cs (hunk 12 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs:314]: - Using existing `Dictionaries.Append()`... I had wanted a more explicit fail-fast behaviour (as per comments in conversation below), if duplicate names attempted to be supplied. Hence not re-using existing. (Also prefered not to modify the passed arguments but return merged)
- Or move `MergeHeaders()` and add tests...  Sure I will look at doing that soon.

File: .github/workflows/build-test.yml (hunk 7 above)
Comment (dvsekhvalnov) [reply .github/workflows/build-test.yml:1]: nice, thanks !

File: jose-jwt/jwe/JWE.cs (hunk 3 above)
Comment (dvsekhvalnov) [reply jose-jwt/jwe/JWE.cs]: THanks !

File: jose-jwt/jose-jwt.csproj (hunk 9 above)
Comment (dvsekhvalnov) [reply jose-jwt/jose-jwt.csproj]: i can test .net40, no problem.

File: jose-jwt/jwe/JWE.cs (hunk 12 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs]: I've just pushed a change to move it to Dictionaries and added unit tests.

File: jose-jwt/jwe/JWE.cs (hunk 13 above)
Comment (dvsekhvalnov) [reply jose-jwt/jwe/JWE.cs]: Isn't it testing `cty` header ?

File: jose-jwt/jwe/JWE.cs (hunk 12 above)
Comment (dvsekhvalnov) [reply jose-jwt/jwe/JWE.cs]: fair enough. Unit tests and we cool with it.

File: jose-jwt/jwe/JWE.cs (hunk 13 above)
Comment (gjermystreeva) [reply jose-jwt/jwe/JWE.cs]: It is already is checking override for `enc` and `alg`... if you take another look it's an xunit `Theory` test rather than a `Fact`... so it is actually running 3 unit tests separately for the 3 supplied InlineData injectedHeaderName names.  `alg` and `enc` (which can never be overridden), and also for `example.com:extra_recipient_header`, which in the test example has already been set for the recipient so can't be set again in extras. If you try running you'll see the 3 different runs.

```
 [Theory]
 [InlineData("example.com:extra_recipient_header")]
 [InlineData("alg")]
 [InlineData("enc")]

```

Comment (dvsekhvalnov) [reply jose-jwt/jwe/JWE.cs]: ok, my bad.

File: UnitTests/DictionariesTest.cs [hunk 14]
```diff
@@ -37,5 +38,74 @@ public void AppendNull()
...
+        [Fact]
+        public void MergeHeaders_MergeNonUniqueKeys_ThrowsArgumentException()
+        {
+            //given
+            var dict1 = new Dictionary<string, object> { { "one", "1" }, { "two", "2" } };
+            var dict2 = new Dictionary<string, object> { { "three", "3" }, { "four", "4" } };
+            var dict3 = new Dictionary<string, object> { { "five", 5 }, { "one", 6 } };
+
+            //when
+            var exception = Record.Exception(() => Dictionaries.MergeHeaders(dict1, dict2, dict3));
```
Comment (dvsekhvalnov) [inline UnitTests/DictionariesTest.cs:103]: Have question here, it seems `.MergeHeaders()` will throw exception if duplicate keys found?  

It seems to be different from current behavior with `enc`, `alg` headers we discussed before. It current duplicates are simply ignored without failing.

Comment (gjermystreeva) [reply UnitTests/DictionariesTest.cs:103]: Hi - yes you are correct the behaviour has changed to be a throw rather than a silent ignore.
Personally I preferred the throw rather than user not realising their request was ignored.
However if you want the previous behaviour maintained I could re-work it to have the previous behaviour? Should I make it do a silent ignore?

Comment (dvsekhvalnov) [reply UnitTests/DictionariesTest.cs:103]: Well, I'd may be prefer to throw it too, but on the other side there are existing users where given change can unexpectedly break applications. So if you don't mind to update it, would appreciate if we can keep compatible behavior.

Comment (gjermystreeva) [reply UnitTests/DictionariesTest.cs:103]: Hi, sorry even though I said I'd change the behaviour I timed-out getting a backward compatible behaviour back in, within time was allowed.
It's more complicated with 3 (or 4) different sources of headers now.
(Note for example Golang errors on dups... https://github.com/latchset/jwcrypto/blob/master/jwcrypto/jwe.py)

Do you think the change in behaviour is a deal-breaker for the PR?

Comment (dvsekhvalnov) [reply UnitTests/DictionariesTest.cs:103]: No, i don't think it is show stopper. Let me think for a while and i get back to you (tomorrowish)

Comment (dvsekhvalnov) [reply UnitTests/DictionariesTest.cs:103]: Ok, probably let's push it with new semantics, can you please add notes to `breaking changes` section here: https://github.com/dvsekhvalnov/jose-jwt#important-upgrade-notes ?

Comment (gjermystreeva) [reply UnitTests/DictionariesTest.cs:103]: Sure - I have put in a note there.

Comment (dvsekhvalnov) [reply UnitTests/DictionariesTest.cs:103]: Looks good to me. Thanks.

File: jose-jwt/jose-jwt.csproj (hunk 9 above)
Comment (dvsekhvalnov) [reply jose-jwt/jose-jwt.csproj]: @gjermystreeva, looks like final one and we can merge it.

**Comment by gjermystreeva** [comment]
Hi @dvsekhvalnov - I made an additional change commit 64e74e4... removed need to pass serialization mode from JWE.Decrypt as it makes it easier for end user I think.

File: jose-jwt/jose-jwt.csproj (hunk 9 above)
Comment (gjermystreeva) [reply jose-jwt/jose-jwt.csproj]: .net40 added back now

**Comment by dvsekhvalnov** [comment]
Ok, i think time to merge it. Thank you so much for contribution!

I'll spend some time testing cross-compatibility and make sure no regressions, as well as try to test new JWE JSON serialization before publishing release. Stay tuned.
